[왓챠챠챠] 2. 모델링 branch 이름 수정 및 models 구현

andy-archive opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [ ] 리뷰 반영
- [ ] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- 3가지 앱 생성 및 각각에 대한 모델링 구현

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
1. branch 이름 수정 (feature/films-models → feature/models)
2. models 구현 (films, users, comments)

(1차 커밋)
1. django app 'films' 생성
2. films/models.py를 ERD에 맞게 구현 완료
3. makemigrations, migrate로 mysql 데이터베이스 연동
4. mysql로 show tables; 하여 테이블 생성 확인

(2차 커밋)
1. 일부 모델링에 'null=True' 적용 (films/models.py)

(3차 커밋)
1. django app 'users' 생성
2. django app 'comments' 생성
3. users/models.py를 ERD에 맞게 구현 완료
4. comments/models.py를 ERD에 맞게 구현 완료
5. makemigrations, migrate로 mysql 데이터베이스 연동
6. mysql로 show tables; 하여 테이블 생성 확인

(4차 커밋)
1. django app 'core' 생성
2. TimeStampModel 생성 (core/models.py)
3. ManyToManyField에 related_name 추가
4. AUTH_USER_MODEL 제거
5. 불필요한 주석 제서 (films/models.py)


<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
- datetimefield에 default=''는 허용되지 않았던 점
- migrations 파일에서 차례대로 검사하기에, 기존 것이 문제가 생기면 새 버전이 아닌 기존 버전을 먼저 수정해야 되는 점

<br />

## :: 기타 질문 및 특이 사항
- soft delete를 하려면, 'deleted_at'이 있는 class에  'is_delete'를 추가로 넣어서 boolean 타입으로, 실제 삭제가 되었는지/아닌지를 판별하는 게 좋을까요?

andy-archive committed:
Add:    films     (django app)
Update: models.py (films/models.py)
Add:    migration (films/migrations)

andy-archive committed:
Update: 'null=True' (films/models.py)

andy-archive committed:
Add: users and comments (django apps)
Update: models.py of users & comments (django apps)

Kang-kyunghun commented on comments/models.py:
```diff
@@ -0,0 +1,37 @@
+from django.db import models
+from django.forms import CharField
```
불필요한 package가 Import 된 것 같습니다. 제거해 주세요.
```suggestion
from django.db import models
```

Kang-kyunghun commented on comments/models.py:
```diff
@@ -0,0 +1,37 @@
+    contents   = models.CharField(max_length=500)
+    deleted_at = models.DateTimeField()
+    films      = models.ForeignKey(Film, on_delete=models.CASCADE)
+    users      = models.ForeignKey(User, on_delete=models.CASCADE)
```
필드명은 특별한 경우를 제외하고는 단수로 해주세요.
(M2M인 경우 복수로 합니다.)

```suggestion
    content    = models.CharField(max_length=500)
    deleted_at = models.DateTimeField()
    film       = models.ForeignKey(Film, on_delete=models.CASCADE)
    user       = models.ForeignKey(User, on_delete=models.CASCADE)
```

Kang-kyunghun commented on comments/models.py:
```diff
@@ -0,0 +1,37 @@
+
+class CommentLike(models.Model):
+    comments = models.ForeignKey('Comment', on_delete=models.CASCADE)
+    users    = models.ForeignKey(User, on_delete=models.CASCADE)
```
```suggestion
    comment  = models.ForeignKey('Comment', on_delete=models.CASCADE)
    user     = models.ForeignKey(User, on_delete=models.CASCADE)
```

Kang-kyunghun commented on comments/models.py:
```diff
@@ -0,0 +1,37 @@
+    contents   = models.CharField(max_length=500)
+    deleted_at = models.DateTimeField()
+    comments   = models.ForeignKey('Comment', on_delete=models.CASCADE)
+    users      = models.ForeignKey(User, on_delete=models.CASCADE)
```
```suggestion
    content    = models.CharField(max_length=500)
    deleted_at = models.DateTimeField()
    comment    = models.ForeignKey('Comment', on_delete=models.CASCADE)
    user       = models.ForeignKey(User, on_delete=models.CASCADE)
```

Kang-kyunghun commented on films/models.py:
```diff
@@ -0,0 +1,99 @@
+from platform import release
+from django.db import models
+
+# 영화 테이블
```
- 불필요한 package 제거
- 불필요한 주석 제거
```suggestion
from django.db import models
```

Kang-kyunghun commented on films/models.py:
```diff
@@ -0,0 +1,99 @@
+    image_url        = models.URLField()
+    running_time_min = models.CharField(max_length=20, default='')
+    descriptions     = models.CharField(max_length=1000, default='')
+    rating_systems   = models.ForeignKey('RatingSystem', on_delete=models.CASCADE)
```
```suggestion
    name             = models.CharField(max_length=100)
    release_date     = models.DateField()
    image_url        = models.URLField()
    running_time_min = models.CharField(max_length=20, default='')
    description      = models.CharField(max_length=1000, default='')
    rating_system    = models.ForeignKey('RatingSystem', on_delete=models.CASCADE)
```

Kang-kyunghun commented on films/models.py:
```diff
@@ -0,0 +1,99 @@
+    class Meta:
+        db_table = 'films'
+
+# 일대다 관계 테이블
```
```suggestion

```

Kang-kyunghun commented on films/models.py:
```diff
@@ -0,0 +1,99 @@
+        db_table = 'rating_systems'
+
+# 다대다 관계 테이블
+
```
```suggestion
```

Kang-kyunghun commented on films/models.py:
```diff
@@ -0,0 +1,99 @@
+
+class Genre(models.Model):
+    name  = models.CharField(max_length=50)
+    films = models.ManyToManyField(Film, through='FilmGenre')
```
M2M field에 related_name을 추가해 주시면 좋을 것 같습니다.
```suggestion
    films = models.ManyToManyField(Film, through='FilmGenre', related_name='genres')
```

Kang-kyunghun commented on watcha_classic/urls.py:
```diff
@@ -13,7 +13,10 @@
 urlpatterns = [
+    # path('films', include('films.urls')),
+    # path('users', include('users.urls')),
+    # path('comments', include('comments.urls')),
```
일반적으로 PR에는 해당 기능에 대한 코드만 있고, 일시적으로 주석 처리된 코드는 안 올리게 됩니다.
해당 주석은 제거해 주세요.

Kang-kyunghun commented on watcha_classic/settings.py:
```diff
@@ -148,3 +151,6 @@
 #슬래시(/) 관련 에러 제거
 APPEND_SLASH = False
+
+# 사용자 지정 모델 사용
+AUTH_USER_MODEL = 'accounts.User'
```
해당 설정을 추가한 이유를 코멘트로 남겨주세요.

miracle-21 replied:
migrate 시 발생한 error를 해결하기 위한 시도 중 하나로, 결과적으로 해결 방법이 아니었어서 삭제하겠습니다.

Kang-kyunghun commented on users/models.py:
```diff
@@ -0,0 +1,34 @@
+class Rate(models.Model):
+    score = models.DecimalField(max_digits=2, decimal_places=1)
+    films = models.ForeignKey(Film, on_delete=models.CASCADE)
+    users = models.ForeignKey('User', on_delete=models.CASCADE)
```
```suggestion
    film  = models.ForeignKey(Film, on_delete=models.CASCADE)
    user  = models.ForeignKey('User', on_delete=models.CASCADE)
```

Kang-kyunghun commented on users/models.py:
```diff
@@ -0,0 +1,34 @@
+
+class WatchList(models.Model):
+    films      = models.ForeignKey(Film, on_delete=models.CASCADE)
+    users      = models.ForeignKey('User', on_delete=models.CASCADE)
```
```suggestion
    film = models.ForeignKey(Film, on_delete=models.CASCADE)
    user = models.ForeignKey('User', on_delete=models.CASCADE)
```

Kang-kyunghun commented on users/models.py:
```diff
@@ -0,0 +1,34 @@
+
+from films.models import Film
+
+class TimeStampModel(models.Model):
+    created_at  = models.DateTimeField(auto_now_add = True)
+    updated_at  = models.DateTimeField(auto_now = True)
+
+    class Meta:
+        abstract = True
```
TimeStampModel을 여러 app에서 사용하고 있습니다.
core app을 만들어 공통으로 사용 되는 것들을 모아 두면 좋을 것 같습니다.
- core app 생성
- core.models.py에 TimeStampModel에 작성
- core.models.py에서 TimeStampModel import 해서 사용

andy-archive committed:
Add:    core                                         (django app)
Update: TimeStampModel                               (core/models.py)
Update: related_name in ManyToManyField              (films/models.py)
Update: unified field name from plural to singular   (films/models.py)
Delete: AUTH_USER_MODEL                              (watcha_classic/settings.py)
Delete: needless comments                            (films/models.py)
